src/librustc_lint/array_into_iter.rs

```diff
@@ -0,0 +1,82 @@
+use crate::lint::{LateContext, LateLintPass, LintArray, LintContext, LintPass};
+use rustc::{
+    lint::FutureIncompatibleInfo,
+    hir,
+    ty,
+};
+use syntax::{
+    errors::Applicability,
+    symbol::{Symbol, sym},
+};
+
+
+declare_lint! {
+    pub ARRAY_INTO_ITER,
+    Warn,
+    "detects calling `into_iter` on arrays",
+    @future_incompatible = FutureIncompatibleInfo {
+        // FIXME: update with real issue
+        reference: "issue #0 <https://github.com/rust-lang/rust/issues/0>",
+        edition: None,
+    };
+}
+
+declare_lint_pass!(
+    /// Checks for instances of calling `into_iter` on arrays.
+    ArrayIntoIter => [ARRAY_INTO_ITER]
+);
+
+impl<'a, 'tcx> LateLintPass<'a, 'tcx> for ArrayIntoIter {
+    fn check_expr(&mut self, cx: &LateContext<'a, 'tcx>, expr: &'tcx hir::Expr) {
+        // We only care about method call expressions.
+        if let hir::ExprKind::MethodCall(call, span, args) = &expr.kind {
+            if call.ident.name.as_str() != "into_iter" {
+                return;
+            }
+
+            // Test if the `self` type is an array type. Early return if not.
+            match cx.tables.expr_ty_adjusted(&args[0]).kind {
+                ty::Ref(_, ty::TyS { kind: ty::Array(..), .. }, _) => {}
+                _ => return,
+            }
+
+            // Check if the method call actually calls the libcore
+            // `IntoIterator::into_iter`. Early return if not.
+            let def_id = cx.tables.type_dependent_def_id(expr.hir_id).unwrap();
+            let trait_id = match cx.tcx.trait_of_item(def_id) {
+                Some(id) => id,
+                None => return,
+            };
+
+            // We compare with the current definition path (which is not the
+            // public path). If the def path changes (because someone moves
+            // some files around), the UI test will fail and this path should
+            // be adjusted.
+            let into_iter_path = [
```

Comment:
This should use `cx.tcx.is_diagnostic_item(sym::IntoIterator, trait_id)` with `#[rustc_diagnostic_item = "IntoIterator"]` added to the definition of `IntoIterator`